module/purpose/Soccer/src/Soccer.cpp

```diff
@@ -124,8 +162,131 @@ namespace module::purpose {
+            return info.robot_id == incoming_robot_id;
         });
+
+        if (it != active_robots.end()) {
+            // if the robot is in active_robots, update timestamp
+            it->last_heard_from = now;
+            // add more data here
+        } else {
+            // Add new robot's info, and re-assign positions
+            RobotInfo new_robot{incoming_robot_id, now};
+            add_robot(new_robot);
+            give_directions();
+        }
+
+        follow_directions(robocup, incoming_robot_id);
+    };
+
+    // find which robot should be our striker
+    uint8_t Soccer::find_striker() {
+        return 0;
+        // TODO: ideas- furthest ahead on field could be striker, or closest to ball
     }
 
+    // give directions if I am the leader
+    void Soccer::give_directions() {
+        bool self_is_leader = active_robots.front().robot_id == PLAYER_ID;
+
+        if (self_is_leader) {
+            auto purposes_msg = std::make_unique<Purposes>();
+            uint8_t striker_idx = find_striker();
+
+            // assign own position
+            self_direct(striker_idx);
+
+            // decide soccer positions
+            for (int i = 0; i < int(active_robots.size()); ++i) {
+                if (i == striker_idx) {
+                    purposes_msg->purposes.push_back({active_robots[i].robot_id, SoccerPosition::STRIKER});
+                } else {
+                    purposes_msg->purposes.push_back({active_robots[i].robot_id, SoccerPosition::DEFENDER});
+                }
+            }
+
+            emit(purposes_msg);
+        }
+    };
+
+    void Soccer::self_direct(uint8_t striker_idx) {
+        if (striker_idx == 0) {
+            // note: at the moment our leader is always our striker, may not always be the case going forward
+            emit<Task>(std::make_unique<Striker>(cfg.force_playing));
+            log<NUClear::DEBUG>("leader made striker");
+        }
+        else {
+            emit<Task>(std::make_unique<Defender>(cfg.force_playing));
+            log<NUClear::DEBUG>("leader made defender");
+        }
+    }
+
+    // listen to directions if they are the leader
+    void Soccer::follow_directions(const RoboCup& robocup, const uint8_t incoming_robot_id) {
+        // A robot may emit leader messages mistakenly. Only listen to real leader.
+        bool other_is_leader = active_robots.front().robot_id == incoming_robot_id;
+        bool purposes_exist = !robocup.purposes.purposes.empty();
+
+        if (other_is_leader && purposes_exist) {
+            for (const auto& purpose : robocup.purposes.purposes) {
+                if (purpose.player_id == PLAYER_ID ) {
+                    if (purpose.purpose == SoccerPosition::DEFENDER) {
+                        emit<Task>(std::make_unique<Defender>(cfg.force_playing));
+                        log<NUClear::DEBUG>("made defender");
```

Comment:
I like these logs and I'd recommend making them `INFO` level so we can see when the robot changes position

Reply:
The only issue is that they'd be "made defender" every 2 secs but I could wrap a guard to not have the task if they already are?

Reply:
ah I see yeah that'd be good, I think it'd be useful to see the change